File: elixir/bowling/bowling.exs
defmodule Bowling do

  defstruct rolls: [], frame: [], frames: 0

  @doc """
    Creates a new game of bowling that can be used to store the results of
    the game
  """

  @spec start() :: any
  def start do
    %Bowling{}
  end

  @doc """
    Records the number of pins knocked down on a single roll. Returns `:ok`
    unless there is something wrong with the given number of pins, in which
    case it returns a helpful message.
  """

  @spec roll(any, integer) :: any | String.t
  def roll(_game, roll) when (roll < 0) or (roll > 10) do
    {:error, "Pins must have a value from 0 to 10"}
  end

  def roll(game = %{frame: []}, roll) when roll == 10 do
    %{ game | rolls: [10 | game.rolls], frames: game.frames+1 }
  end

  def roll(game = %{frame: []}, roll) do
    %{ game | rolls: [roll | game.rolls], frame: [roll] }
  end

  def roll(%{frame: [prev]}, roll) when prev+roll > 10 do
    {:error, "Pin count exceeds pins on the lane"}
  end

  def roll(game, roll) do
    %{ game | rolls: [roll | game.rolls], frame: [], frames: game.frames+1 }
  end

  @doc """
    Returns the score of a given game of bowling if the game is complete.
    If the game isn't complete, it returns a helpful message.
  """

  @spec score(any) :: integer | String.t
  def score(%{rolls: rolls, frames: frames}) do
    cond do
      frames < 10 -> {:error, "Score cannot be taken until the end of the game"}
      true -> rolls |> Enum.reverse |> do_score(1,0)
    end
  end

  defp do_score([], _frame, score), do: score
  defp do_score(_rolls, 11, score), do: score

  defp do_score([10|tail], frame, score) do
    [r2,r3] = Enum.take(tail, 2)
    do_score(tail, frame+1, score+10+r2+r3)
  end

  defp do_score([r1, r2|tail], frame, score) when r1+r2 == 10 do
    [r3] = Enum.take(tail, 1)
    do_score(tail, frame+1, score + 10 + r3)
  end

  defp do_score([r1, r2|tail], frame, score) do
    do_score(tail, frame+1, score+r1+r2)
  end

end

File: go/bowling/bowling.go
package bowling

import "errors"

type Game struct {
	balls []int
}

func NewGame() *Game {
	return &Game{}
}

func (g *Game) Roll(pins int) error {
	if g.done() {
		return errors.New("Game is done")
	}

	if pins < 0 || pins > 10 {
		return errors.New("Invalid number of pins")
	}

	midFrame := len(g.balls)%2 == 1
	prevPins := 0
	if midFrame {
		prevPins = g.balls[len(g.balls)-1]
	}

	// can't knock down more than 10 pins in a frame
	if midFrame && prevPins+pins > 10 {
		return errors.New("Invalid number of pins for frame")
	}

	g.balls = append(g.balls, pins)

	// add an extra ball if it was a strike
	if pins == 10 {
		g.balls = append(g.balls, 0)
	}

	return nil
}

func (g *Game) Score() (int, error) {
	score := 0

	// can't caculate score until game is done
	if !g.done() {
		return 0, errors.New("Game is not done yet")
	}

	// score each frame
	for t := 0; t < 20; t += 2 {
		first, second := g.balls[t], g.balls[t+1]
		score += first + second

		// if it was a spare, add the next ball
		if first+second == 10 {
			score += g.balls[t+2]
		}

		// if it was a strike, add the 2nd next ball
		if first == 10 {
			// if next frame was a strike, add the first ball from the next+1 frame
			if g.balls[t+2] == 10 {
				score += g.balls[t+4]
			} else {
				score += g.balls[t+3]
			}
		}
	}

	return score, nil
}

func (g *Game) done() bool {
	balls := len(g.balls)

	// still going if we haven't rolled 20 balls yet
	if balls >= 0 && balls < 20 {
		return false
	}

	// if we've hit 20 balls, we are done unless 10th frame was strike or spare
	if balls == 20 {
		return g.balls[18]+g.balls[19] < 10
	}

	// we only get 1 bonus roll if 10th frame was a spare
	if balls == 21 {
		return g.balls[18] < 10
	}

	// we get 2 bonus rolls if 10th frame was a strike and first bonus roll was a strike
	if balls == 22 && g.balls[18] == 10 && g.balls[20] == 10 {
		return false
	}

	return true
}
